File: programs/java/src/lppp/ch05/Tokenizer.java
package lppp.ch05;

import java.io.IOException;

/**
 * Created by Pierre Nugues on 25/07/15.
 */
public class Tokenizer {

    public static void main(String[] args) throws IOException {
        FileReader reader = new FileReader();
        String text = reader.readFile(args[0]);
        Tokenizer tokenizer = new Tokenizer();
        String[] words = tokenizer.tokenize(text);
        for (int i = 0; i < words.length; i++) {
            System.out.println(words[i]);
        }
    }

    String[] tokenize(String text) {
        //String [] words = text.split("[\\s\\-,;:!?.’\'«»()–...&‘’“”*—]+");
        //String [] words = text.split("[^a-zåàâäæçéèêëîïôöœßùûüÿA-ZÅÀÂÄÆÇÉÈÊËÎÏÔÖŒÙÛÜŸ’\\-]+");
        //String [] words = text.split("\\W+"); // Not unicode friendly
        String[] words = text.split("\\P{L}+");
        return words;
    }

    String tokenize2(String text) {
        String words = text.replaceAll("\\P{L}+", " ");
        return words;
    }
}

File: programs/ch05/python/tokenizer.py
"""
Tokenizers
Usage: python tokenizer.py < corpus.txt
"""
__author__ = "Pierre Nugues"

import sys
import regex as re

text = """Tell me, O muse, of that ingenious hero who
travelled far and wide after he had sacked the famous
town of Troy."""


def tokenize(text):
    """uses the nonletters to break the text into words
    returns a list of words"""
    # words = re.split(r'[\s\-,;:!?.’\'«»()–...&‘’“”*—]+', text)
    # words = re.split(r'[^a-zåàâäæçéèêëîïôöœßùûüÿA-ZÅÀÂÄÆÇÉÈÊËÎÏÔÖŒÙÛÜŸ’\-]+', text)
    # words = re.split(r'\W+', text)
    words = re.split(r'\P{L}+', text)
    words.remove('')
    return words


def tokenize2(text):
    """uses the letters to break the text into words
    returns a list of words"""
    # words = re.findall(r'[a-zåàâäæçéèêëîïôöœßùûüÿA-ZÅÀÂÄÆÇÉÈÊËÎÏÔÖŒÙÛÜŸ’\-]+', text)
    # words = re.findall(r'\w+', text)
    words = re.findall(r'\p{L}+', text)
    return words


def tokenize3(text):
    """uses the punctuation and nonletters to break the text into words
    returns a list of words"""
    # text = re.sub(r'[^a-zåàâäæçéèêëîïôöœßùûüÿA-ZÅÀÂÄÆÇÉÈÊËÎÏÔÖŒÙÛÜŸ’'()\-,.?!:;]+', '\n', text)
    # text = re.sub(r'([,.?!:;)('-])', r'\n\1\n', text)
    text = re.sub(r'[^\p{L}\p{P}]+', '\n', text)
    text = re.sub(r'(\p{P})', r'\n\1\n', text)
    text = re.sub(r'\n+', '\n', text)
    return text.split()


def tokenize4(text):
    """uses the punctuation and symbols to break the text into words
    returns a list of words"""
    spaced_tokens = re.sub(r'([\p{S}\p{P}])', r' \1 ', text)
    one_token_per_line = re.sub(r'\s+', '\n', spaced_tokens)
    tokens = one_token_per_line.split()
    return tokens


if __name__ == '__main__':
    text = sys.stdin.read()
    """words = tokenize(text)
    for word in words:
        print(word)
    words = tokenize2(text)
    print(words)"""
    words = tokenize4(text)
    print(words)
